### PR TITLE
Impression tacking

### PR DESCRIPTION
**Pull request by anshumansworld**
This patch includes:

- Impression tracking (impression would be counted if view is visible to user for at least one second in a row)

- It tracks the click of view and opens browser/play store on the basis of click url

- Inform publisher about these events.

### CODE REVIEW

File: library/src/main/java/net/pubnative/library/models/PubnativeAdModel.java [hunk 1]
```diff
@@ -44,4 +138,81 @@ public String getBeacon(String beaconType) {
...
+     * Start tracking of ad view
+     * @param view ad view
+     * @param listener listener for callbacks
+     */
+    public void startTracking(View view, Listener listener) throws NullPointerException {
+
+        if(view == null) {
+            throw new NullPointerException("view can't be null");
+        }
+        startTracking(view, view, listener);
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/models/PubnativeAdModel.java]: We should NEVER raise an exception, the listener is there for that matter.

In other order of things argument checks should be done in the inner method instead of here

File: library/src/main/java/net/pubnative/library/models/PubnativeAdModel.java [hunk 2]
```diff
@@ -44,4 +138,81 @@ public String getBeacon(String beaconType) {
...
+
+        Log.v(TAG, "startTracking(view:" + view + ", clickableView:" + clickableView + ", listener:" + listener + ")");
+
+        if(view == null) {
+            throw new NullPointerException("view can't be null");
+        }
+
+        if(clickableView == null) {
+            throw new NullPointerException("clickable view can't be null");
+        }
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/models/PubnativeAdModel.java]: no throw in the entire library, drop the call and log an error or callback with exception

File: library/src/main/java/net/pubnative/library/models/PubnativeAdModel.java (hunk 1 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/models/PubnativeAdModel.java]: :+1:

File: library/src/main/java/net/pubnative/library/models/PubnativeAdModel.java (hunk 2 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/models/PubnativeAdModel.java]: :+1:

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+    private static String                   TAG                             = PubnativeAdTracker.class.getSimpleName();
+
+    private PubnativeAdModel.Listener       mListener;
+    private View                            mView;
+    private View                            mClickableView;
+    private ViewTreeObserver                mViewTreeObserver;
+    private PubnativeAdModel                mPubnativeAdModel;
+    private final ScheduledExecutorService  mExecutor;
+    private boolean                         isTracked                       = false;
+    private boolean                         isTrackingStopped               = false;
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: why not calling them `mIsTrackingStopped` and `mIsTracked` ?

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+
+                        try {
+                            Thread.sleep(VISIBILITY_CHECK_INTERVAL);
+                        } catch (InterruptedException e) {
+                            e.printStackTrace();
+                        }
+                    }
+
+                }
+            }, VISIBILITY_CHECK_INTERVAL, TimeUnit.MILLISECONDS);
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: Can you comment this runnable inside to know the cases with natural language?

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+        String impressionUrl = mPubnativeAdModel.getBeacon(PubnativeBeacon.BeaconType.IMPRESSION);
+
+        if (TextUtils.isEmpty(impressionUrl)) {
+
+            mHandler.post(new Runnable() {
+                @Override
+                public void run() {
+                    invokeOnImpressionFailed(new MalformedURLException("Can not confirm impression, no Beacon URL found"));
+                }
+            });
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: wouldn't it better to post the handler runnable inside the `invokeOn` methods ?? As I understand this, the invokeOnImpression should ensure It's being executed on the correct thread

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+        Log.v(TAG, "handleClickEvent()");
+
+        if (!TextUtils.isEmpty(mPubnativeAdModel.getClickUrl())) {
+
+            URLOpener urlOpener = new URLOpener(mView.getContext());
+            urlOpener.openInBackground(mPubnativeAdModel.getClickUrl(), true, urlOpenerListener);
+
+        } else {
+
+            invokeOnClickFailed(new MalformedURLException("Can not open ad, no click_url found"));
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: Why not using the handler here? You never know where is the tracking coming from.. but we should execute this on the same thread

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+        }
+    }
+
+    protected URLOpener.Listener urlOpenerListener = new URLOpener.Listener() {
+        @Override
+        public void onURLOpenerStart(String url) {
+
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: When this occurs, I like to set a comment type `// Do nothing`

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+        }
+
+        @Override
+        public void onURLOpenerRedirect(String url) {
+
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: Can we log redirections?

File: library/src/main/java/net/pubnative/library/models/PubnativeAdModel.java [hunk 9]
```diff
@@ -44,4 +138,81 @@ public String getBeacon(String beaconType) {
...
+         * Called when click is confirmed
+         * @param view The view that was clicked
+         */
+        void onPubnativeAdModelClicked(View view);
+
+        /**
+         * Called when error occurred while click
+         * @param exception error
+         */
+        void onPubnativeAdModelClickFailed(Exception exception);
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/models/PubnativeAdModel.java]: For public methods, this callbacks should return the object they are sticked to.. (imagine multiple ads in the same class, they could be interpreted as the same one, for refreshing in example).

So please return the model also in all callbacks as the first argument

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+    }
+
+    private void handleClickEvent() {
+
+        Log.v(TAG, "handleClickEvent()");
+
+        if (!TextUtils.isEmpty(mPubnativeAdModel.getClickUrl())) {
+
+            URLOpener urlOpener = new URLOpener(mView.getContext());
+            urlOpener.openInBackground(mPubnativeAdModel.getClickUrl(), true, urlOpenerListener);
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: I personally prefr to define the URLOpener listener here instead of defining a method or variable.

Whether it's small and can be defined inside, or it's big and then we should assign `this` instead of a variable

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java (hunk 5 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: :+1:

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java (hunk 6 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: :+1:

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java (hunk 8 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: it's already there in URLOpener

File: library/src/main/java/net/pubnative/library/models/PubnativeAdModel.java (hunk 9 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/models/PubnativeAdModel.java]: :+1:

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java (hunk 3 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: :+1:

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java (hunk 8 above)
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: :+1:

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java (hunk 4 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: :+1:

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java (hunk 7 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: :+1:

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java (hunk 10 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: :+1:

File: library/src/test/java/net/pubnative/library/models/PubnativeAdModelTest.java [hunk 11]
```diff
@@ -63,10 +72,87 @@ public void testGetBeaconWitContainedValueReturnsBeacon() {
...
+        model.startTracking(adView, listener);
+    }
+
+    @Test
+    public void testStartTrackingWithNullListener() {
+
+        PubnativeAdModel            model       = spy(PubnativeAdModel.class);
+
+        Activity activity = Robolectric.buildActivity(Activity.class).create().get();
+        View adView = new View(activity);
```
Comment (cerberillo) [inline library/src/test/java/net/pubnative/library/models/PubnativeAdModelTest.java]: Robolectric has a ShadowView class for this :)

File: library/src/test/java/net/pubnative/library/tracking/PubnativeAdTrackerTest.java [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+@Config(constants = BuildConfig.class,
+        sdk = 21)
+public class PubnativeAdTrackerTest {
+
+    @Test
+    public void testImpressionFailureForInvalidImpressionURL() {
+
+        PubnativeAdModel            model       = spy(PubnativeAdModel.class);
+        PubnativeAdModel.Listener   listener    = spy(PubnativeAdModel.Listener.class);
+        PubnativeBeacon             beacon      = spy(PubnativeBeacon.class);
```
Comment (cerberillo) [inline library/src/test/java/net/pubnative/library/tracking/PubnativeAdTrackerTest.java]: Instead of a spy, you can use a regular object here

File: library/src/test/java/net/pubnative/library/tracking/PubnativeAdTrackerTest.java [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+
+        List<PubnativeBeacon> beacons = new ArrayList<PubnativeBeacon>();
+        beacons.add(beacon);
+
+        model.setBeacons(beacons);
+
+        Activity activity = Robolectric.buildActivity(Activity.class).create().get();
+        View adView = new View(activity);
+
+        PubnativeAdTracker tracker = new PubnativeAdTracker(adView, adView, listener, model);
```
Comment (cerberillo) [inline library/src/test/java/net/pubnative/library/tracking/PubnativeAdTrackerTest.java]: here you should use a spy

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+
+        } else {
+
+            invokeOnClickFailed(new MalformedURLException("Can not open ad, no click_url found"));
+        }
+    }
+
+    private void invokeOnImpressionFailed(final Exception exception) {
+
+        if(mListener != null) {
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: the listener check should be done inside the runnable, otherwise, there could be some type of post>release>invoke on null.

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java [hunk 15]
```diff
@@ -0,0 +1,282 @@
...
+
+                    // note first visible time
+                    long firstVisibleTime = System.currentTimeMillis() - VISIBILITY_CHECK_INTERVAL;
+
+                    // loop to make sure view is visible on screen for at least 1sec
+                    while(System.currentTimeMillis() - firstVisibleTime < VISIBILITY_TIME_THRESHOLD + VISIBILITY_CHECK_INTERVAL) {
+
+                        // If view is already tracked or not visible it returns from the loop without confirming impression
+                        if(mIsTracked || !SystemUtils.isVisibleOnScreen(mView, VISIBILITY_PERCENTAGE_THRESHOLD)) {
+                            return;
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: In this case, you should continue checking whenever the view becomes visible again..

Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: no need for that, we have globallayoutlistener for that on ViewTreeObserver

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java (hunk 14 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: :+1:

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java (hunk 15 above)
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: then why you need `!SystemUtils.isVisibleOnScreen(mView, VISIBILITY_PERCENTAGE_THRESHOLD)) {`?

Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: right, it is to check the percentage and visibility. This method is being invoked whenever there is change in layout or scroll.

Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: :+1:

File: library/src/main/java/net/pubnative/library/models/PubnativeAdModel.java [hunk 16]
```diff
@@ -44,4 +138,109 @@ public String getBeacon(String beaconType) {
...
     }
+
+    /**
+     * Interface definition for callbacks to be invoked when impression confirmed/failed, ad clicked/clickfailed
+     */
+    public interface Listener {
+
+        /**
+         * Called when impression is confirmed
+         * @param view The view where impression confirmed
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/models/PubnativeAdModel.java]: Check spelling "the view was impression confirmed"

File: library/src/main/java/net/pubnative/library/models/PubnativeAdModel.java [hunk 17]
```diff
@@ -44,4 +138,109 @@ public String getBeacon(String beaconType) {
...
+
+            }
+
+            @Override
+            public void onClickFailed(Exception exception) {
+
+                if(listener != null) {
+
+                    listener.onPubnativeAdModelClickFailed(PubnativeAdModel.this, exception);
+                }
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/models/PubnativeAdModel.java]: Some changes
* Make the model listener of the ad tracker instead of having such a big amount of code inside the method.
* Make every listener check and invoke inside specific methods for them `invokeImpressionConfirmed` `invokeImpressionFailed` `invokeClickConfirmed` `invokeClickFailed` so you can also create independent tests and verifies for the model-tracker callback connection

File: library/src/main/java/net/pubnative/library/network/PubnativeAPIRequestTask.java [hunk 18]
```diff
@@ -66,12 +66,16 @@ private void executeRequest() {
...
+                throw new Exception("Server error: " + responseCode);
             }
         } catch (IOException e) {
 
             Log.e(TAG, e.getMessage());
             pubnativeAPIResponse.setError(e);
+        } catch (Exception e) {
+
+            Log.e(TAG, e.getMessage());
+            pubnativeAPIResponse.setError(e);
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/network/PubnativeAPIRequestTask.java]: `Exception` covers `IOException` and since `IOException` is not doing anything special, we should remove `IOException` catch case

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java [hunk 19]
```diff
@@ -0,0 +1,310 @@
...
+        startTracking();
+    }
+
+    public void stopTracking() {
+
+        Log.v(TAG, "stopTracking()");
+
+        mExecutor.shutdownNow();
+        mListener = null;
+        mIsTrackingStopped = true;
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: You should also reset `mIsTracking`

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java [hunk 20]
```diff
@@ -0,0 +1,310 @@
...
+    private static String                   TAG                             = PubnativeAdTracker.class.getSimpleName();
+
+    private Listener                        mListener;
+    private View                            mView;
+    private View                            mClickableView;
+    private ViewTreeObserver                mViewTreeObserver;
+    private final ScheduledExecutorService  mExecutor;
+    private boolean                         mIsTracked                      = false;
+    private boolean                         mIsTracking                     = false;
+    private boolean                         mIsTrackingStopped              = false;
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: mIsTracking should cover mIsTRackingStopped case, isn't this variable duplicated?

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java [hunk 21]
```diff
@@ -0,0 +1,310 @@
...
+
+    private ViewTreeObserver.OnScrollChangedListener onScrollChangedListener = new ViewTreeObserver.OnScrollChangedListener() {
+        @Override
+        public void onScrollChanged() {
+            checkImpression();
+        }
+    };
+
+    private void checkImpression() {
+        if (mIsTracking || mIsTracked || mIsTrackingStopped || mExecutor.isShutdown()) {
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: IsTracked & a live-tracking variable should be enough, 4 variables is way too much

File: library/src/main/java/net/pubnative/library/models/PubnativeAdModel.java [hunk 22]
```diff
@@ -44,4 +138,109 @@ public String getBeacon(String beaconType) {
...
+     * start tracking of your ad view
+     * @param view ad view
+     * @param clickableView clickable view
+     * @param listener listener for callbacks
+     */
+    public void startTracking(final View view, final View clickableView, final Listener listener) {
+
+        Log.v(TAG, "startTracking(view:" + view + ", clickableView:" + clickableView + ", listener:" + listener + ")");
+
+        mPubnativeAdTracker = new PubnativeAdTracker(view, clickableView, getBeacon(PubnativeBeacon.BeaconType.IMPRESSION), getClickUrl(), new PubnativeAdTracker.Listener(){
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/models/PubnativeAdModel.java]: You should check if there is already a tracker attached. In order to be able to reset this, then you should reset the tracker when you start a new request

File: library/src/main/java/net/pubnative/library/utils/SystemUtils.java [hunk 23]
```diff
@@ -94,4 +98,68 @@ public static Location getLastLocation(Context context) {
...
+        int viewWidth = view.getWidth();
+        int viewHeight = view.getHeight();
+
+        int bottomRightX = topLeftX + viewWidth;
+        int bottomRightY = topLeftY + viewHeight;
+
+        // Create Rect with view's current position
+        Rect viewRect = new Rect(topLeftX, topLeftY, bottomRightX, bottomRightY);
+        // This will store the rect which come after clipping the viewRect with the screen
+        Rect intersectionRect = new Rect();
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/utils/SystemUtils.java]: Let's define this somewhere more useful, like inside next if sentence

File: library/src/main/java/net/pubnative/library/utils/SystemUtils.java [hunk 24]
```diff
@@ -94,4 +98,68 @@ public static Location getLastLocation(Context context) {
...
+                if (percentageVisible >= percentage) {
+
+                    return true;
+                }
+            }
+        }
+
+        // CASE 4: When the view is outside of screen bounds such that no part of it is visible then the default value(false) of result will return.
+        // We don't need to put explicit condition for this case.
+        return false;
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/utils/SystemUtils.java]: Let's please create a `result` variable at the beggining, so readability don't gets interrupted

method{
  var result = DEFAULT_VALUE;
  // DO WHATEVER YOU NEED ASSIGNING RESULT
  return result;
}

File: library/src/test/java/net/pubnative/library/models/PubnativeAdModelTest.java [hunk 25]
```diff
@@ -63,10 +81,79 @@ public void testGetBeaconWitContainedValueReturnsBeacon() {
...
+    public void testStartTrackingWithClickableView() {
+
+        PubnativeAdModel            model       = spy(PubnativeAdModel.class);
+        PubnativeAdModel.Listener   listener    = mock(PubnativeAdModel.Listener.class);
+        ImageView                   adView      = spy(new ImageView(applicationContext));
+        ImageView                   clickableView = spy(new ImageView(applicationContext));
+
+        model.startTracking(adView, clickableView, listener);
+
+        verify(clickableView, times(1)).setOnClickListener(any(View.OnClickListener.class));
```
Comment (cerberillo) [inline library/src/test/java/net/pubnative/library/models/PubnativeAdModelTest.java]: you should also check that this is never done over the adView

File: library/src/test/java/net/pubnative/library/models/PubnativeAdModelTest.java [hunk 26]
```diff
@@ -63,10 +81,79 @@ public void testGetBeaconWitContainedValueReturnsBeacon() {
...
+        model.startTracking(adView, null);
+    }
+
+    @Test
+    public void testStartTrackingWithClickableView() {
+
+        PubnativeAdModel            model       = spy(PubnativeAdModel.class);
+        PubnativeAdModel.Listener   listener    = mock(PubnativeAdModel.Listener.class);
+        ImageView                   adView      = spy(new ImageView(applicationContext));
+        ImageView                   clickableView = spy(new ImageView(applicationContext));
```
Comment (cerberillo) [inline library/src/test/java/net/pubnative/library/models/PubnativeAdModelTest.java]: Can you use View instead of ImageView?

File: library/src/test/java/net/pubnative/library/models/PubnativeAdModelTest.java [hunk 27]
```diff
@@ -63,10 +81,79 @@ public void testGetBeaconWitContainedValueReturnsBeacon() {
...
+    public void testStartTrackingWithNullClickableView() {
+
+        PubnativeAdModel            model       = spy(PubnativeAdModel.class);
+        PubnativeAdModel.Listener   listener    = mock(PubnativeAdModel.Listener.class);
+
+        ImageView                   adView      = spy(new ImageView(applicationContext));
+
+        model.startTracking(adView, null, listener);
+        verify(listener, times(1)).onPubnativeAdModelClickFailed(eq(model), any(Exception.class));
+    }
```
Comment (cerberillo) [inline library/src/test/java/net/pubnative/library/models/PubnativeAdModelTest.java]: I'm missing listener callbacks tests

File: library/src/test/java/net/pubnative/library/tracking/PubnativeAdTrackerTest.java [hunk 28]
```diff
@@ -0,0 +1,91 @@
...
+
+        PubnativeAdTracker.Listener listener    = mock(PubnativeAdTracker.Listener.class);
+        ImageView                   adView      = spy(new ImageView(applicationContext));
+
+        PubnativeAdTracker          tracker     = spy(new PubnativeAdTracker(adView, adView, "", "", listener));
+
+        tracker.invokeOnResponse("");
+
+        verify(listener, times(1)).onImpressionConfirmed();
+    }
```
Comment (cerberillo) [inline library/src/test/java/net/pubnative/library/tracking/PubnativeAdTrackerTest.java]: This tests are not over the impressionSuccess, but over the callbacks, please rename to give a meaningful name

**Comment by cerberillo** [comment]
In order to make tests more modular, I would tests every behaviour where it belongs. In this way we should change tests in the following way.

* Model tests should check that startTRacking of the model starts tracking of the adTracker
* Tracker tests should check that startTracking attach listener to clickable view

In this way we have more modular tests and not so class-crossed

File: library/src/main/java/net/pubnative/library/network/PubnativeAPIRequestTask.java (hunk 18 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/network/PubnativeAPIRequestTask.java]: :+1:

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java (hunk 19 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: :+1:

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java (hunk 20 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: the name is eventually similar but use is different. I will rename variable.
:+1:

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java (hunk 21 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: will try to optimize.
:+1:

File: library/src/main/java/net/pubnative/library/models/PubnativeAdModel.java (hunk 22 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/models/PubnativeAdModel.java]: :+1:

File: library/src/main/java/net/pubnative/library/utils/SystemUtils.java (hunk 23 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/utils/SystemUtils.java]: :+1:

File: library/src/main/java/net/pubnative/library/utils/SystemUtils.java (hunk 24 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/utils/SystemUtils.java]: :+1:

File: library/src/test/java/net/pubnative/library/models/PubnativeAdModelTest.java (hunk 25 above)
Comment (anshumansworld) [inline library/src/test/java/net/pubnative/library/models/PubnativeAdModelTest.java]: :+1:

File: library/src/test/java/net/pubnative/library/models/PubnativeAdModelTest.java (hunk 26 above)
Comment (anshumansworld) [inline library/src/test/java/net/pubnative/library/models/PubnativeAdModelTest.java]: :+1:

File: library/src/test/java/net/pubnative/library/tracking/PubnativeAdTrackerTest.java (hunk 28 above)
Comment (anshumansworld) [inline library/src/test/java/net/pubnative/library/tracking/PubnativeAdTrackerTest.java]: :+1:

File: library/src/main/java/net/pubnative/library/models/PubnativeAdModel.java (hunk 17 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/models/PubnativeAdModel.java]: :+1:

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java [hunk 29]
```diff
@@ -0,0 +1,311 @@
...
+                    while (System.currentTimeMillis() - firstVisibleTime < VISIBILITY_TIME_THRESHOLD + VISIBILITY_CHECK_INTERVAL) {
+
+                        Log.v(TAG, "checkImpression(), within time threshold checking. Current time is: " + System.currentTimeMillis());
+
+                        // If view is already tracked or not visible it returns from the loop without confirming impression
+                        if (mIsTracked || !SystemUtils.isVisibleOnScreen(mView, VISIBILITY_PERCENTAGE_THRESHOLD)) {
+
+                            Log.v(TAG, "checkImpression(), either already tracked or not visible anymore. Already tracked is: " + mIsTracked + " & Current time is: " + System.currentTimeMillis());
+
+                            mIsTrackingInProgress = false;
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: what about this?
```java
mViewTreeObserver.removeGlobalOnLayoutListener(onGlobalLayoutListener);
mViewTreeObserver.removeOnScrollChangedListener(onScrollChangedListener);
mExecutor.shutdownNow();
```

I guess It would make sense to set up a stop method that reset the tracker state and ensures that whenever that method is called nothing will happen from that moment in advance

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java [hunk 30]
```diff
@@ -0,0 +1,311 @@
...
+        }
+    }
+
+    protected void startImpressionRequest() {
+
+        Log.v(TAG, "startImpressionRequest()");
+
+        if(mTrackingShouldStop) {
+            return;
+        }
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: I guess that we've just checked trackingShouldStop in the same thread (inside the loop), we can get rid of this edge case (overcomplicates)

File: library/src/test/java/net/pubnative/library/models/PubnativeAdModelTest.java [hunk 31]
```diff
@@ -63,10 +81,93 @@ public void testGetBeaconWitContainedValueReturnsBeacon() {
...
     }
+
+    @Test
+    public void testStartTrackingWithValidListener() {
+
+        PubnativeAdModel            model       = spy(PubnativeAdModel.class);
+        PubnativeAdModel.Listener   listener    = mock(PubnativeAdModel.Listener.class);
+        View                        adView      = spy(new View(applicationContext));
+
+        model.startTracking(adView, listener);
```
Comment (cerberillo) [inline library/src/test/java/net/pubnative/library/models/PubnativeAdModelTest.java]: you shouldn't use startTracking to set up the listener, you should set it up directly

Comment (cerberillo) [inline library/src/test/java/net/pubnative/library/models/PubnativeAdModelTest.java]: Same for all other tests

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java (hunk 30 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: :+1:

File: library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java (hunk 29 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/tracking/PubnativeAdTracker.java]: :+1:

File: library/src/test/java/net/pubnative/library/models/PubnativeAdModelTest.java (hunk 31 above)
Comment (anshumansworld) [inline library/src/test/java/net/pubnative/library/models/PubnativeAdModelTest.java]: :+1:

File: library/src/test/java/net/pubnative/library/tracking/PubnativeAdTrackerTest.java [hunk 32]
```diff
@@ -0,0 +1,113 @@
...
+    @Test
+    public void testInValidClickListener() {
+
+        View                        adView          = spy(new View(applicationContext));
+        View                        clickableView   = spy(new View(applicationContext));
+
+        PubnativeAdTracker          tracker         = spy(new PubnativeAdTracker(adView, clickableView, "", "", null));
+
+        verify(adView, never()).setOnClickListener(any(View.OnClickListener.class));
+    }
```
Comment (cerberillo) [inline library/src/test/java/net/pubnative/library/tracking/PubnativeAdTrackerTest.java]: I miss tracker `invoke's` methods tests (ensure to do it asynchronously, because it requires handler to execute).

File: library/src/test/java/net/pubnative/library/tracking/PubnativeAdTrackerTest.java [hunk 33]
```diff
@@ -0,0 +1,113 @@
...
+    public void testImpressionSuccess() {
+
+        PubnativeAdTracker.Listener listener    = mock(PubnativeAdTracker.Listener.class);
+        View                        adView      = spy(new View(applicationContext));
+
+        PubnativeAdTracker          tracker     = spy(new PubnativeAdTracker(adView, adView, "", "", listener));
+
+        tracker.invokeOnResponse("");
+
+        verify(listener, times(1)).onImpressionConfirmed(adView);
```
Comment (cerberillo) [inline library/src/test/java/net/pubnative/library/tracking/PubnativeAdTrackerTest.java]: This is implementation confirmation, you should check invokeOnImpressionConfirmed directly instead of doing it through invokeOnResponse (another class callback)

File: library/src/test/java/net/pubnative/library/tracking/PubnativeAdTrackerTest.java (hunk 32 above)
Comment (cerberillo) [inline library/src/test/java/net/pubnative/library/tracking/PubnativeAdTrackerTest.java]: You should also check some `new PubnativeAdTracker()` inits with null  adView arguments to avoid crashing (and check listener callback)

Comment (anshumansworld) [inline library/src/test/java/net/pubnative/library/tracking/PubnativeAdTrackerTest.java]: :+1:

**Comment by cerberillo** [comment]
Looks good, let's merge :+1:
